### PR TITLE
Patch with accelerate xpu

### PR DESCRIPTION
**Pull request by abhilash1910**
Patch for Accelerate XPU support.
cc @sgugger @muellerzr

### CODE REVIEW

**Comment by abhilash1910** [comment]
@sgugger  seems like ruff is flagging errors on other files (which are not edited in this commit ). Could you take a look ? Thanks

**Comment by sgugger** [comment]
@abhilash1910 I am no longer working at Hugging Face, so you should ping @muellerzr to help :-)

**Review by muellerzr — APPROVED** [review]
Thanks! One small nit, and let's ensure tests fully pass before we merge. (via rebasing etc)

File: tests/trainer/test_trainer_distributed.py [hunk 1]
```diff
@@ -159,6 +160,20 @@ def test_trainer(self):
         # successful return here == success - any errors would have caused an error in the sub-call
 
 
+class TestTrainerDistributedXPU(TestCasePlus):
+    @require_torch_multi_xpu
```
[the remark this answers is not in the document]
Comment (abhilash1910) [reply tests/trainer/test_trainer_distributed.py]: Thanks for the suggestion, could you re-review?

File: tests/models/markuplm/test_processor_markuplm.py [hunk 2]
```diff
@@ -340,7 +340,7 @@ def test_processor_case_3(self):
...
...6, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216], [216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216], [216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216], [216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216], [216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216, 216]]  # noqa:
```
Comment (amyeroberts) [inline tests/models/markuplm/test_processor_markuplm.py:343]: There shouldn't be any change here

File: src/transformers/trainer_utils.py [hunk 3]
```diff
@@ -510,14 +528,24 @@ def stop(self, stage):
...
+                }
+            elif is_torch_xpu_available():
+                self.gpu_mem_used_now = self.torch.xpu.memory_allocated()
+                self.gpu_mem_used_peak = self.torch.xpu.max_memory_allocated()
+                self.gpu[self.cur_stage] = {
+                    "begin": self.gpu_mem_used_at_start,
+                    "end": self.gpu_mem_used_now,
+                    "alloc": (self.gpu_mem_used_now - self.gpu_mem_used_at_start),
+                    "peaked": max(0, self.gpu_mem_used_peak - self.gpu_mem_used_now),
+                }
```
Comment (amyeroberts) [inline src/transformers/trainer_utils.py:548]: We can reduce some of this repeated code 

```suggestion
            if torch.cuda.is_available():
                self.gpu_mem_used_now = self.torch.cuda.memory_allocated()
                self.gpu_mem_used_peak = self.torch.cuda.max_memory_allocated()
            elif is_torch_xpu_available():
                self.gpu_mem_used_now = self.torch.xpu.memory_allocated()
                self.gpu_mem_used_peak = self.torch.xpu.max_memory_allocated()
            else:
                raise ValueError("No available GPU device found!")
                
            self.gpu[self.cur_stage] = {
                "begin": self.gpu_mem_used_at_start,
                "end": self.gpu_mem_used_now,
                "alloc": (self.gpu_mem_used_now - self.gpu_mem_used_at_start),
                "peaked": max(0, self.gpu_mem_used_peak - self.gpu_mem_used_now),
            }
```

@muellerzr  - is it possible for `self.torch` to be set but `torch.cuda.is_available()` to be `False` ?

File: src/transformers/utils/import_utils.py [hunk 4]
```diff
@@ -528,6 +528,26 @@ def get_major_and_minor_from_version(full_version):
...
 
+@lru_cache
+def is_torch_xpu_available(check_device=False):
+    "Checks if `intel_extension_for_pytorch` is installed and potentially if a XPU is in the environment"
+    if is_ipex_available():
+        import torch
+    else:
+        return False
+
+    import intel_extension_for_pytorch  # noqa: F401
```
Comment (amyeroberts) [inline src/transformers/utils/import_utils.py:539]: It's a bit funny to have the condition checking if ipex is available, but then only have `torch` imported in that if/else block

```suggestion
    if not is_ipex_available():
        return False
        
    import torch
    import intel_extension_for_pytorch  # noqa: F401
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this additional functionality! 

Just some small nits on structure. Could you run `make style` and push any changes - this should make the code quality checks pass.

File: tests/models/markuplm/test_processor_markuplm.py (hunk 2 above)
Comment (abhilash1910) [reply tests/models/markuplm/test_processor_markuplm.py:343]: This change was done when I had run 'make style' . It reformatted this segment. I can try to revert it.

File: src/transformers/utils/import_utils.py (hunk 4 above)
Comment (abhilash1910) [reply src/transformers/utils/import_utils.py:539]: Yeah this is something  I was thinking about . Will change

**Comment by abhilash1910** [comment]
@amyeroberts thanks for the suggestions. However when I try to run  ```make style``` then many unrelated files are getting edited and ruff test triggers failure on some unmodified files.  One such example was the markupml model test script (make style changed the formatting).Could you suggest how to go about this ?

**Comment by amyeroberts** [comment]
@abhilash1910 Are you running the ruff and black versions expected in the library? They can be installed using `pip install -Ue .[quality]`

File: src/transformers/utils/import_utils.py [hunk 5]
```diff
@@ -29,6 +29,7 @@
 from types import ModuleType
 from typing import Any, Tuple, Union
 
+import torch
```
Comment (amyeroberts) [inline src/transformers/utils/import_utils.py:32]: You can't import torch here, it has to be safely imported within the functions

File: src/transformers/utils/import_utils.py [hunk 6]
```diff
@@ -528,6 +529,24 @@ def get_major_and_minor_from_version(full_version):
     return True
 
 
+@lru_cache
+def is_torch_xpu_available(check_device=False):
+    "Checks if `intel_extension_for_pytorch` is installed and potentially if a XPU is in the environment"
+    if not is_ipex_available():
+        return False
+
+    import intel_extension_for_pytorch  # noqa: F401
```
Comment (amyeroberts) [inline src/transformers/utils/import_utils.py:538]:
```suggestion
    import torch
    import intel_extension_for_pytorch  # noqa: F401
```

File: tests/models/markuplm/test_processor_markuplm.py (hunk 2 above)
Comment (amyeroberts) [reply tests/models/markuplm/test_processor_markuplm.py:343]: You should be able to revert by running `git checkout -- tests/models/markuplm/test_processor_markuplm.py`

File: src/transformers/training_args.py [hunk 7]
```diff
@@ -1356,11 +1357,17 @@ def __post_init__(self):
...
-                    raise ValueError(
-                        "Your setup doesn't support bf16/gpu. You need torch>=1.10, using Ampere GPU with cuda>=11.0"
-                    )
+                elif not self.use_cpu:
+                    if torch.cuda.is_available() and not is_torch_bf16_gpu_available():
+                        # gpu
+                        raise ValueError(
+                            "Your setup doesn't support bf16/gpu. You need torch>=1.10, using Ampere GPU with cuda>=11.0"
+                        )
+                    elif not is_torch_xpu_available():
```
Comment (amyeroberts) [inline src/transformers/training_args.py:1366]: This check doesn't match the exception raised below as the torch version isn't checked: it needs to be added to `is_torch_xpu_available`. 

If bf16 supported in XPU?

**Review by amyeroberts — COMMENTED** [review]
Left a few more comments. The main comment is about importing `torch` at the top level of import_utils.py - this is the cause of most of the failures in the CI runs at the moment. 

The changes to the unrelated files from `make style` is very odd 🤔  Only files which have had changes committed should be linted (and these files I'd expect to already be in the correct format as they're recent). Are any of these files open in your IDE? Could you share the versions of black and ruff in your environment?

File: src/transformers/training_args.py (hunk 7 above)
Comment (abhilash1910) [reply src/transformers/training_args.py:1366]: bf16 is supported in xpu.

File: src/transformers/utils/import_utils.py (hunk 5 above)
Comment (abhilash1910) [reply src/transformers/utils/import_utils.py]: Thanks for this suggestion

**Comment by abhilash1910** [comment]
I think the initial tests pass now; I have not added the style fixes of the other mentioned files ( I used pip install -Ue .[quality] for black/ruff/isort etc ) . Seems when I run make style on markuplm , it is formatting seq list (without it tests fail on make style).
Could you trigger the slow tests ? Thanks

**Comment by abhilash1910** [comment]
@amyeroberts @muellerzr  could you please re-review (retrigger tests)? Since this integration is a little urgent for us ; thanks again for all suggestions.

File: src/transformers/training_args.py [hunk 8]
```diff
@@ -1357,11 +1358,22 @@ def __post_init__(self):
...
+                        # xpu
+                        import torch
+
+                        parsed_torch_version_base = version.parse(version.parse(torch.__version__).base_version)
+                        is_torch_greater_or_equal_than_1_12 = parsed_torch_version_base >= version.parse("1.12")
```
Comment (amyeroberts) [inline src/transformers/training_args.py:1372]: `is_torch_greater_or_equal_than_1_12` is already defined in pytorch_utils and should be imported from there

```suggestion
```

**Review by amyeroberts — APPROVED** [review]
Thanks for adding! 

Looks good to me. Just one small comment. Once that's resolved we're good to merge!

**Review by amyeroberts — APPROVED** [review]
LGTM - thanks for iterating and adding this to the library!

**Review by muellerzr — APPROVED** [review]
Thanks again!

**Pull request merged**